flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java

```diff
@@ -46,8 +47,23 @@ static CheckpointInProgressRequest completeOutput(long checkpointId) {
 		return new CheckpointInProgressRequest("completeOutput", checkpointId, ChannelStateCheckpointWriter::completeOutput, false);
 	}
 
-	static ChannelStateWriteRequest write(long checkpointId, InputChannelInfo info, Buffer... flinkBuffers) {
-		return new CheckpointInProgressRequest("writeInput", checkpointId, writer -> writer.writeInput(info, flinkBuffers), recycle(flinkBuffers), false);
+	static ChannelStateWriteRequest write(long checkpointId, InputChannelInfo info, CloseableIterator<Buffer> iterator) {
+		return new CheckpointInProgressRequest(
+			"writeInput",
+			checkpointId,
+			writer -> {
+				try {
+					while (iterator.hasNext()) {
+						Buffer buffer = iterator.next();
+						checkArgument(buffer.isBuffer());
+						writer.writeInput(info, buffer);
+					}
+				} finally {
+					iterator.close();
+				}
+			},
+			throwable -> iterator.close(),
```

Comment:
are you sure this won't be called twice? Or if it will, that it will not double release the buffers/files?